### PR TITLE
[server] Restart PVC workspace from latest valid backup

### LINKED ISSUE #14030: [PVC] Workspace fails on restart with misconfigured initializer if some prior instance "failed"
context: https://gitpod.slack.com/archives/C047BR51GDR

especially thread: https://gitpod.slack.com/archives/C047BR51GDR/p1666257320068649

example workspaces:
 1. gitpodio-gitpod-njdud87zz0n https://gitpod.io/admin/workspaces/gitpodio-gitpod-njdud87zz0n
 2. gitpodio-gitpod-0crlhy9mhdk https://gitpod.io/admin/workspaces/gitpodio-gitpod-0crlhy9mhdk

example trace for 1. containing wrong initializer: https://jaeger-eu-prod.workspace-preview.gitpod-io-dev.com/trace/71c1f5f3e97f870a

Code:
 1. [here](https://github.com/gitpod-io/gitpod/blob/a4870ce5db87d35a4547afee07836e82018e50e3/components/server/src/workspace/workspace-starter.ts#L339-L347) we set `hasValidBackup` to false if any prior instance `failed`
 2. [here](https://github.com/gitpod-io/gitpod/blob/a4870ce5db87d35a4547afee07836e82018e50e3/components/server/src/workspace/workspace-starter.ts#L1543-L1551) volumeSnapshots is `undefined`
 3. so we end up [here](https://github.com/gitpod-io/gitpod/blob/a4870ce5db87d35a4547afee07836e82018e50e3/components/server/src/workspace/workspace-starter.ts#L1718-L1724), where we create a BackupInitializer, but without setting `FromVolumeSnapshot` to `true` and without setting `VolumeSnapshot`.

/cc @sagor999

### PR DESCRIPTION
**Pull request by geropl**
## Related Issue(s)

Fixes #14030

## How to test
 1. open a workspace on this PR
 1. make sure `user_pvc` is enabled for your user in ConfigCat
     1. after changing the rule, `kubectl rollout restart deployment/server`
 1. [login](https://gpl-14030-pvc.preview.gitpod-dev.com/workspaces)
 4. start a workspace
 5. check the workspace pod, and verify that the label `gitpod.io/pvcFeature: "true"` is present :heavy_check_mark: 
 6. stop the workspace
 7. start and stop that workspace another time
 8. manipulate the DB:
     1. Find the instances of your wrokspacId: `select id, creationTime, status from d_b_workspace_instance WHERE workspaceId = ' ... ';`
     2. Set the latest instance to `failed`; `UPDATE d_b_workspace_instance SET status = JSON_SET(status, "$.conditions.failed", 'test') WHERE id = ' ... ';`
 9. Start your workspace again, and see how it work :heavy_check_mark: 

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

**Comment by kylos101** [comment]
:wave: @geropl when testing, I recommend this at least: `/werft run with-integration-tests=workspace with-large-vm=true`

**Comment by geropl** [comment]
> I recommend this at least: /werft run with-integration-tests=workspace with-large-vm=true

Hey  @kylos101 , thx for the pointer!
Can you explain why those'd help? :thinking:

1. this should not affect any workspace integration tests
2. what does `with-large-vm` do? :-o

**Comment by kylos101** [comment]
Howdy @geropl ,

> this should not affect any workspace integration tests

Good point! :facepalm: 

> what does with-large-vm do? :-o

This helps tests "run faster", sometimes they can timeout, by giving the preview environment more resources.

**Comment by jankeromnes** [comment]
> when testing, I recommend this at least: `/werft run with-integration-tests=workspace with-large-vm=true`

FYI, it could help to put these options at the bottom of the PR description, so that every build of this PR gets them. 👍

File: components/server/src/workspace/workspace-starter.ts [hunk 1]
```diff
@@ -336,14 +336,14 @@ export class WorkspaceStarter {
...
-            const hasValidBackup = pastInstances.some(
-                (i) => !!i.status && !!i.status.conditions && !i.status.conditions.failed,
-            );
             let lastValidWorkspaceInstance: WorkspaceInstance | undefined;
-            if (hasValidBackup) {
-                lastValidWorkspaceInstance = pastInstances.reduce((previousValue, currentValue) =>
-                    currentValue.creationTime > previousValue.creationTime ? currentValue : previousValue,
-                );
+            // Sorted from latest to oldest
+            for (const i of pastInstances.sort((a, b) => (a.creationTime > b.creationTime ? -1 : 1))) {
```
Comment (easyCZ) [inline components/server/src/workspace/workspace-starter.ts:341]: Would we want to sort on `stopped` time? A backup takes place after a workspace stops which means that we're looking for stopped workspaces, not creation time, no?

**Comment by jankeromnes** [comment]
FYI, I tested this PR with various scenarii that failed in the context of Gitpod's PVC prebuilds / #13991 :
- incremental prebuilds
- same prebuild different branch
- opening prebuilds directly

(I have no idea if my workspaces/prebuilds were actually using PVC, but) doing these things worked perfectly. 🙌

File: components/server/src/workspace/workspace-starter.ts (hunk 1 above)
Comment (geropl) [reply components/server/src/workspace/workspace-starter.ts:341]: Instances are totally ordered, so it should not make a difference, except that creationTime might be a little less brittle.

Comment (easyCZ) [reply components/server/src/workspace/workspace-starter.ts:341]: Ah, it's an instance, not a workspace. Thanks.

**Comment by geropl** [comment]
> (I have no idea if my workspaces/prebuilds were actually using PVC

see [description at the top](https://github.com/gitpod-io/gitpod/pull/14053#issue-1416883247) on how to configure and check.

**Review by sagor999 — APPROVED** [review]
LGTM

**Review by easyCZ — APPROVED** [review]

**Pull request merged**